_posts/2020-07-21-cube-functionality-to-kartothek.markdown

```diff
@@ -0,0 +1,299 @@
+1         4                   4.0   Dresden      DE 2018-01-02  51.050407  13.737262
+2         6                   6.0   Hamburg      DE 2018-01-01  53.551086   9.993682
+3         5                   4.0   Hamburg      DE 2018-01-02  53.551086   9.993682
+4         6                   6.0    London      UK 2018-01-01        NaN        NaN
+5         8                   4.0    London      UK 2018-01-02        NaN        NaN
+6        20                   NaN  Santiago      CL 2018-01-01        NaN        NaN
+7        22                   NaN  Santiago      CL 2018-01-02        NaN        NaN 
+```
+
+You can also **Delete** entire datasets (or the entire cube).
+
+```python
+>>> from klee2.io.eager import delete_cube
+>>> datasets_still_in_cube = delete_cube(
+...     cube=cube,
+...     store=store,
+...     datasets=["transformed"],
+... )
+>>> query_cube(
+...     cube=cube,
+...     store=store,
+... )[0]
+   avg_temp      city country        day   latitude  longitude
+0         8   Dresden      DE 2018-01-01  51.050407  13.737262
+1         4   Dresden      DE 2018-01-02  51.050407  13.737262
+2         6   Hamburg      DE 2018-01-01  53.551086   9.993682
+3         5   Hamburg      DE 2018-01-02  53.551086   9.993682
+4         6    London      UK 2018-01-01        NaN        NaN
+5         8    London      UK 2018-01-02        NaN        NaN
+6        20  Santiago      CL 2018-01-01        NaN        NaN
+7        22  Santiago      CL 2018-01-02        NaN        NaN
+```
+
+## Additional Features of Cube
+
+*	**Multiple-datasets**: When mapping multiple parts (tables or datasets) to Kartothek, using multiple datasets allows users to copy, backup and delete them separately. 
+Index structures are bound to datasets.
+*	**Seed-Based Join System / Partition-alignment**: When data is stored in multiple parts (tables or datasets), the question is how to expose it to the user during read operations.
+ Seed based Join marks a single part as seed which provides seed dataset in the cube, all other parts are just additional columns.
+ Cube uses lazy approach of seed based join, 
+ since it better supports independent copies and backups of datasets and also simplifies some of our processing pipelines (e.g. geolocation data can blindly be fetched for too many locations and dates.)	
+
+
+## Command Line Interface (CLI)
+Kartothek features a **command line interface (CLI)** for some cube operations. 
+To use it, create a  **skv.yml** file that describes [storefact](https://github.com/JDASoftwareGroup/storefact) stores and use commands to gather information of the required cube.
+
+Sample skv.yml:
+```
+dataset:
+   type: hfs
+   path: path/to/data
+```
+
+Here we use **geodata** cube to get some information.
+```shell
+>>>kartothek_cube geodata info  (gives geodata cube info)
+>>>kartothek_cube geodata stats  (for cube scan)
+>>>kartothek_cube --help  (To get list of  ktk_cube commands)
+```
+
+## Outlook
+TODO:What features are coming up next ??
```

Comment:
```markdown
- **API cleanup:** The API surface of kartothek grow organically over the years and we plan to re-design it. While doing so, we will incorporate our learnings regarding API design and will also prune some features that are not needed anymore or that did not match their expectations (e.g. the original multi-table design).
- **Ecosystem integration:** At this point in time, there are multiple dataset formats (e.g. [Apache Arrow](https://arrow.apache.org/docs/python/dataset.html), [Apache Iceberg](https://iceberg.apache.org/), [Delta Lake](https://delta.io/)) and we will investigate how to evolve kartothek as a library and as a format to align better with the ecosystem and enable new features (like schema migrations and time travel), while providing the stability and safety that our users require
- **Query Planning:** Currently the kartothek query planner solely relies on file-level information (file names for primary indices and separate index files for secondary indices). It would be great to also use the RowGroup-level statistics as specified in [Apache Parquet](https://github.com/apache/parquet-format) to improve query performance. We will have a look at [the work Dask already did](https://github.com/dask/dask/blob/55445565c3746f97f2bc50d5628a484576cef90e/dask/dataframe/io/parquet/core.py) in this area.
```

Reply:
Added this content to outlook.